src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,130 @@
+		var latestVersion = new Version( GithubUtilities.GetReleaseVersion( releaseData ) );
+		if( latestVersion <= localVersion ) {
+			Console.WriteLine( $"Already own the latest version {latestVersion}" );
+			return;
+		}
+		string archiveName = GetOSFileName();
+		string downloadUrl = releaseData?.Assets?.FirstOrDefault( a => a.Name == archiveName )?.BrowserDownloadUrl
+			?? string.Empty;
+		if( string.IsNullOrWhiteSpace( downloadUrl ) ) {
+			throw new BmxException( "Failed to get update download url" );
+		}
+
+		string downloadPath = Path.GetTempFileName();
+
+		string? currentFilePath = Environment.ProcessPath;
+		if( string.IsNullOrEmpty( currentFilePath ) ) {
+			throw new BmxException( "BMX could not update" );
+		}
+		string currentDirectory = Path.GetDirectoryName( currentFilePath )!;
+		string backupPath = $"{BmxPaths.OLD_BMX_VERSION_FILE_NAME}v{localVersion}.old.bak";
+
+		try {
+			File.Move( currentFilePath, backupPath );
+		} catch( IOException ex ) {
+			throw new BmxException( "Could move the old version, try with elevated permissions", ex );
+		} catch {
+			throw new BmxException( "BMX could not update" );
+		}
+
+		try {
+			var archiveRes = await httpClient.GetAsync( downloadUrl, HttpCompletionOption.ResponseHeadersRead );
+			using( var fs = new FileStream( downloadPath, FileMode.Create, FileAccess.Write, FileShare.None ) ) {
+				await archiveRes.Content.CopyToAsync( fs );
+				await fs.FlushAsync();
+				fs.Dispose();
+			}
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentFilePath );
+			throw new BmxException( "Failed to download the update", ex );
+		}
+
+		try {
+			string extension = Path.GetExtension( downloadUrl ).ToLowerInvariant();
+
+			if( extension.Equals( ".zip" ) ) {
+				ExtractZipFile( downloadPath, currentDirectory );
+			} else if( extension.Equals( ".gz" ) ) {
+				ExtractTarGzipFile( downloadPath, currentDirectory );
+			} else {
+				throw new Exception( "Unknown archive type" );
+			}
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentFilePath );
+			throw new BmxException( "Failed to update with new files", ex );
+		} finally {
+			File.Delete( downloadPath );
+		}
+	}
+
+	private static string GetOSFileName() {
+
+		if( RuntimeInformation.IsOSPlatform( OSPlatform.Windows ) ) {
+			return "bmx-win-x64.zip";
+		} else if( RuntimeInformation.IsOSPlatform( OSPlatform.OSX ) ) {
+			return "bmx-osx-x64.tar.gz";
+		} else if( RuntimeInformation.IsOSPlatform( OSPlatform.Linux ) ) {
+			return "bmx-linux-x64.tar.gz";
+		} else {
+			throw new Exception( "Unknown OS" );
+		}
+	}
+
+	public static void Cleanup() {
+		string processDirectory = Path.GetDirectoryName( BmxPaths.OLD_BMX_VERSION_FILE_NAME ) ?? string.Empty;
+		if( string.IsNullOrEmpty( processDirectory ) ) {
+			return;
+		}
+		foreach( string file in Directory.GetFiles( processDirectory, "*.old.bak" ) ) {
+			try {
+				File.Delete( file );
+			} catch( Exception ex ) {
```

Comment:
as per build warning
```suggestion
			} catch( Exception ) {
```